### ISSUE TITLE
adaface loss

### ISSUE THREAD

**Issue by eeric** [deepinsight/insightface#2007]
could author add adaface loss to your insightface?
adaface code: https://github.com/mk-minchul/AdaFace/blob/master/head.py

**Comment by jacqueline-weng** [comment]
I have incorporated in my code and runs well. However, I do not have resource to test its performance. It would be appreciated if anyone could test it.


```
class AdaFace(torch.nn.Module):
    def __init__(self,
                 m=0.4,
                 h=0.333,
                 s=64.,
                 t_alpha=1.0,
                 ):
        super(AdaFace, self).__init__()
        self.m = m 
        self.eps = 1e-3
        self.h = h
        self.s = s
        self.t_alpha = t_alpha
        self.register_buffer('t', torch.zeros(1))
        self.register_buffer('batch_mean', torch.ones(1)*(20))
        self.register_buffer('batch_std', torch.ones(1)*100)

        print('\n\AdaFace with the following property')
        print('self.m', self.m)
        print('self.h', self.h)
        print('self.s', self.s)
        print('self.t_alpha', self.t_alpha)


    def forward(self, embeddings, labels, weight_activated=None, fp16=None):

        with torch.cuda.amp.autocast(fp16):
            norm_embeddings = normalize(embeddings)
            norm_weight_activated = normalize(weight_activated)
            logits = linear(norm_embeddings, norm_weight_activated)
        if fp16:
            logits = logits.float()

        logits = logits.clamp(-1,1)

        norms = torch.norm(embeddings, 2, -1, keepdim=True)

        safe_norms = torch.clip(norms, min=0.001, max=100) # for stability
        safe_norms = safe_norms.clone().detach()

        # update batchmean batchstd
        with torch.no_grad():
            mean = safe_norms.mean().detach()
            std = safe_norms.std().detach()
            self.batch_mean = mean * self.t_alpha + (1 - self.t_alpha) * self.batch_mean
            self.batch_std =  std * self.t_alpha + (1 - self.t_alpha) * self.batch_std

        margin_scaler = (safe_norms - self.batch_mean) / (self.batch_std+self.eps) # 66% between -1, 1
        margin_scaler = margin_scaler * self.h # 68% between -0.333 ,0.333 when h:0.333
        margin_scaler = torch.clip(margin_scaler, -1, 1) # normalized ||zi|| 
        index = torch.where(labels != -1)[0]
        margin_scaler = margin_scaler[index,:]
        # g_angular
        g_angular = self.m * margin_scaler * -1
        cos_g_angular = torch.cos(g_angular)
        sin_g_angular = torch.sin(g_angular)
        # g_add
        g_add = self.m + (self.m * margin_scaler)

        # when theta != y_i
        
        target_logit = logits[index, labels[index].view(-1)]
        sin_theta = torch.sqrt(1.0 - torch.pow(target_logit, 2))
        final_target_logit = target_logit.unsqueeze(dim=1) * cos_g_angular - sin_theta.unsqueeze(dim=1) * sin_g_angular - g_add / self.s
        logits[index, labels[index].view(-1)] = final_target_logit.squeeze()
        # when theta == y_i
        logits = logits * self.s

        return logits

```

**Comment by eeric** [comment]
@noreply 
your code was right in accordance with Arcface in insightface style.

**Comment by xlflove** [comment]
Hello, has the implementation of AdaFace been replicated under InsightFace?

**Comment by TaharChettaoui** [comment]
Hello,

I found a compatible implementation of AdaFace in CVLface:
https://github.com/mk-minchul/CVLface/blob/main/cvlface/research/recognition/code/run_v1/classifiers/partial_fc/partial_fc.py
